### ISSUE TITLE
Provide step-by-step instructions to compile and install Protobuf

### ISSUE THREAD

**Issue by coryan** [protocolbuffers/protobuf#12787]
**What language does this apply to?**

C++

**Describe the problem you are trying to solve.**

Developers trying to install Protobuf from source may miss critical dependencies and/or fail to notice that some dependencies need to be installed too.  The top-level [instructions](https://github.com/protocolbuffers/protobuf#protobuf-runtime-installation) redirect to the [src directory](https://github.com/protocolbuffers/protobuf#protobuf-runtime-installation) which redirect to the [cmake](https://github.com/protocolbuffers/protobuf/blob/main/cmake/README.md) directory.

The instructions there are a bit scattered.  They mention Abseil being installed in passing. And they mention `-Dprotobuf_ABSL_PROVIDER=package` only for Windows.  The instructions for Linux are near [the bottom](https://github.com/protocolbuffers/protobuf/blob/main/cmake/README.md#linux-builds) and one needs to skip stuff about running unit tests, getting submodules, DLLs, compiler warnings, etc.

It is too easy for developers on other platforms to miss the Abseil flags, and get confused using Abseil as a submodule.

**Describe the solution you'd like**

I think you should break your build instructions into separate use-cases.  The instructions to compile Protobuf as a contributor (with the intention of running tests and sending patches) should be separate from the instructions to compile the code as a consumer (with the intention of just using Protobuf).

For example, the how-to guides for contributors could start with:

```
**ATTENTION:** These are instructions to compile Protobuf as a contributor to the project.
If you want to use Abseil in your own project use [these instructions](/path/to/doc.md) instead.
```

while the instructions for consumers of the library could start with:

```
**ATTENTION:** These are instructions to compile Protobuf as a consumer of the libraries.
If you want to contribute to the project use [these instructions](/path/to/doc.md) instead.
```

The instructions for each platform should be complete.  Ideally tested, I have used Docker scripts to test the instructions and then treat the Dockerfiles as the source of truth to ensure they stay in sync.  They should say something like:

```
## Linux

To compile and install Protobuf on Linux you must have Abseil installed.  Follow [Abseil's instructions](link here) to do so.

Then compile the project with

    cmake -S . -B .build -D...<all necessary flags here>
    cmake --build .build
    cmake --install .build

## Windows

<repeat for windows with whatever variations>

## macOS

<again>
```

Only once you have completed the instructions for the easy case you may want to discuss variations (using `ZLIB`, using DLLs, etc.)

**Describe alternatives you've considered**

**Additional context**

N/A

**Comment by coryan** [comment]
AFAICT this is still not resolved.

**Comment by Logofile** [comment]
That's correct. The inactive label was applied by a bot, but I still have this on my to-do list.

**Comment by Logofile** [comment]
I still plan to work on this.

**Comment by amlamarra** [comment]
I hope this doesn't stay closed. It would be really nice to have updated documentation. I am now currently having issues with any version of Protobuf that requires Abseil (>= v22.x). My project requires Protobuf but I've been getting linker errors with "undefined reference to" messages. I'm sure I'm doing something wrong, but WHAT? Anyway, I had to switch to using Protobuf v21.12.

**Comment by Logofile** [comment]
Sorry. I missed that this got closed. I agree that this should stay open.

**Comment by coryan** [comment]
@amlamarra maybe this? https://github.com/protocolbuffers/protobuf/issues/12292#issuecomment-1529680040

**Comment by konstantin1722** [comment]
Still have a lot of `undefined reference to` absl lib messages. I tried several solutions from topic #12292, including building different versions of abseil and using different protobuf build parameters. But some of them seem odd in terms of the obviousness of the solution and they all didn't work for me.

**Bottom line is there a unified solution on how to build this on linux?** From past threads, I take it the current documentation is not correct?

**Comment by amlamarra** [comment]
> @amlamarra maybe this? [#12292 (comment)](https://github.com/protocolbuffers/protobuf/issues/12292#issuecomment-1529680040)

Yes, I definitely needed that. But also, I was using static libs from abseil and needed to add the libs for the call to CMake's `target_link_library()` in the correct order.

**Comment by Logofile** [comment]
Commenting to keep this issue open.

**Comment by Logofile** [comment]
This should remain open. I don't know when the updates will be made.
